File: main_project/message_pipe/src/cpp/Producer.cpp

#include "Producer.h"

namespace xce{
namespace messagepipe{

bool Producer::send(const std::string& zk_address, const MessageDataPtr& producerData) {
	if(producerData->getCluster().empty() || producerData->getSubject().empty() || producerData->getData().empty()){
		return false;
	}
	NodeSetPtr nodes = ClusterManager::getInstance()->getConnection(zk_address,producerData->getCluster());
	if (nodes != NULL) {
		return nodes->send(producerData);
	} else {
		return false;
	}
}

void Producer::sendBatch(const std::string& zk_address, const std::vector<MessageDataPtr>& producerDatas) {
	for (std::vector<MessageDataPtr>::const_iterator it=producerDatas.begin();it!=producerDatas.end();++it) {
		send(zk_address, *it);
	}
}

}
}

File: main_project/message_pipe/src/java/src/main/java/com/renren/messagepipe/producer/Producer.java
package com.renren.messagepipe.producer;

import java.io.PrintStream;
import java.util.TimeZone;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import com.google.inject.Guice;
import com.google.inject.Injector;
import com.renren.messagepipe.common.BrokerConnection;
import com.renren.messagepipe.common.ClusterManager;
import com.yammer.metrics.Metrics;
import com.yammer.metrics.core.MetricsRegistry;
import com.yammer.metrics.core.TimerMetric;
import com.yammer.metrics.guice.InstrumentationModule;
import com.yammer.metrics.guice.Timed;
import com.yammer.metrics.reporting.ConsoleReporter;
import com.yammer.metrics.util.MetricPredicate;

public class Producer {

	private static Producer instance = new Producer();
	private final TimerMetric timer = Metrics.newTimer(getClass(),
			"directory-listing", TimeUnit.MILLISECONDS, TimeUnit.SECONDS);
	private static MetricsRegistry registry;
	{
		// Injector injector = Guice.createInjector(new
		// InstrumentationModule());
		// instance = injector.getInstance(Producer.class);
		// registry = injector.getInstance(MetricsRegistry.class);
		// ConsoleReporter.enable(10, TimeUnit.SECONDS);
	}

	public static Producer getInstance() {
		return instance;
	}

	// @Timed
	public boolean send(final MessageData producerData) {
		final BrokerConnection brokerConnection = ClusterManager
				.getConnection(producerData.getCluster());
		if (brokerConnection != null) {
			return brokerConnection.send(producerData);
		} else {
			return false;
		}
	}

	public void send(java.util.List<MessageData> producerDatas) {
		for (MessageData data : producerDatas) {
			send(data);
		}
	}
}
